infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientServerRouter.java

```diff
@@ -0,0 +1,605 @@
+        if (isConnectionInitiator(session)) {
+            if(sessionManager.getIncomingSessions().contains(session)) {
+                if (sessionToSinkVerifyLeadership.get(session).getRemoteLeaderNodeId().isPresent()) {
+                    clientChannelAdapter.send(sessionToSinkVerifyLeadership.get(session).getRemoteLeaderNodeId().get(), response);
+                } else {
+                    sessionToSinkVerifyLeadership.get(session).input(
+                            new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.REMOTE_LEADER_LOSS));
+                }
+            }
+        } else {
+            try {
+                this.serverChannelAdapter.send(response);
+                log.info("Sent response: {}", response);
+            } catch (IllegalArgumentException e) {
+                log.warn("Illegal response type. Ignoring message.", e);
+            }
+        }
+    }
+
+    /**
+     * Receive Corfu Message from the Channel Adapter for further processing
+     *
+     * @param msg received corfu message
+     */
+    @Override
+    public void receive(CorfuMessage.ResponseMsg msg) {
+        try {
+            LogReplicationSession session = msg.getHeader().getSession();
+            String nodeId = msg.getPayload().getLrLeadershipLoss().getNodeId();
+
+            // If it is a Leadership Loss Message re-trigger leadership discovery
+            if (msg.getPayload().getPayloadCase() == CorfuMessage.ResponsePayloadMsg.PayloadCase.LR_LEADERSHIP_LOSS) {
+                if (isOutgoingSession(session)) {
+                    this.sessionToRuntimeFSM.get(session)
+                            .input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.REMOTE_LEADER_LOSS, nodeId));
+                } else {
+                    sessionToSinkVerifyLeadership.get(session)
+                            .input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.REMOTE_LEADER_LOSS, nodeId));
+                }
+                return;
+            }
+
+            if (msg.getPayload().getPayloadCase().equals(CorfuMessage.ResponsePayloadMsg.PayloadCase.LR_SUBSCRIBE_MSG)) {
+                // Start runtimeFSM
+                sessionToRuntimeFSM.get(session).setRemoteLeaderNodeId(nodeId);
+                sessionToRuntimeFSM.get(session).start();
+                log.debug("runtimeFSM started for session {}", session);
+                sessionToRuntimeFSM.get(session)
+                        .input(new LogReplicationRuntimeEvent(LogReplicationRuntimeEvent.LogReplicationRuntimeEventType.ON_CONNECTION_UP, nodeId));
+            } else {
+                // Route the message to the handler.
+                if (log.isTraceEnabled()) {
+                    log.trace("Message routed to {}}: {}", msgHandler.getClass().getSimpleName(), msg);
+                }
+                msgHandler.handleMessage(null, msg, this);
```

Comment:
since we now have new handlers and both Source and Sink use them, it will be good to add an assertion/precondition check for the type of responses received based on the Source or Sink role.

Same with the request messages also.